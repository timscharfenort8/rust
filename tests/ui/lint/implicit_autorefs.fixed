//@ check-pass
//@ run-rustfix

#![allow(dead_code)] // for the rustfix-ed code

use std::mem::ManuallyDrop;
use std::ptr::addr_of_mut;
use std::ptr::addr_of;
use std::ops::Deref;

unsafe fn test_const(ptr: *const [u8]) -> *const [u8] {
    addr_of!((&(*ptr))[..16])
    //~^ WARN implicit auto-ref
}

struct Test {
    field: [u8],
}

unsafe fn test_field(ptr: *const Test) -> *const [u8] {
    let l = (&(*ptr).field).len();
    //~^ WARN implicit auto-ref

    addr_of!((&(*ptr).field)[..l - 1])
    //~^ WARN implicit auto-ref
}

unsafe fn test_builtin_index(a: *mut [String]) {
    _ = (&(*a)[0]).len();
    //~^ WARN implicit auto-ref

    _ = (&(&(*a))[..1][0]).len();
    //~^ WARN implicit auto-ref
    //~^^ WARN implicit auto-ref
}

unsafe fn test_overloaded_deref_const(ptr: *const ManuallyDrop<Test>) {
    _ = addr_of!((&(*ptr)).field);
    //~^ WARN implicit auto-ref
}

unsafe fn test_overloaded_deref_mut(ptr: *mut ManuallyDrop<Test>) {
    _ = addr_of_mut!((&mut (*ptr)).field);
    //~^ WARN implicit auto-ref
}

unsafe fn test_manually_overloaded_deref() {
    struct W<T>(T);

    impl<T> Deref for W<T> {
        type Target = T;
        fn deref(&self) -> &T { &self.0 }
    }

    let w: W<i32> = W(5);
    let w = addr_of!(w);
    let _p: *const i32 = addr_of!(*(&**w));
    //~^ WARN implicit auto-ref
}

unsafe fn test_no_attr(ptr: *mut ManuallyDrop<u8>) {
    ptr.write(ManuallyDrop::new(1)); // should not warn, as `ManuallyDrop::write` is not
                                     // annotated with `#[rustc_no_implicit_auto_ref]`
}

fn main() {}
