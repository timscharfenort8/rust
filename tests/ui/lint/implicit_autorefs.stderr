warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:12:14
   |
LL |     addr_of!((*ptr)[..16])
   |              ^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
   = note: `#[warn(dangerous_implicit_autorefs)]` on by default
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     addr_of!((&(*ptr))[..16])
   |              ++      +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:21:13
   |
LL |     let l = (*ptr).field.len();
   |             ^^^^^^^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     let l = (&(*ptr).field).len();
   |             ++            +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:24:14
   |
LL |     addr_of!((*ptr).field[..l - 1])
   |              ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     addr_of!((&(*ptr).field)[..l - 1])
   |              ++            +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:29:9
   |
LL |     _ = (*a)[0].len();
   |         ^^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     _ = (&(*a)[0]).len();
   |         ++       +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:32:9
   |
LL |     _ = (*a)[..1][0].len();
   |         ^^^^^^^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     _ = (&(*a)[..1][0]).len();
   |         ++            +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:32:9
   |
LL |     _ = (*a)[..1][0].len();
   |         ^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     _ = (&(*a))[..1][0].len();
   |         ++    +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:38:18
   |
LL |     _ = addr_of!((*ptr).field);
   |                  ^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     _ = addr_of!((&(*ptr)).field);
   |                  ++      +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:43:22
   |
LL |     _ = addr_of_mut!((*ptr).field);
   |                      ^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     _ = addr_of_mut!((&mut (*ptr)).field);
   |                      +++++       +

warning: implicit auto-ref creates a reference to a dereference of a raw pointer
  --> $DIR/implicit_autorefs.rs:57:26
   |
LL |     let _p: *const i32 = addr_of!(**w);
   |                          ^^^^^^^^^^^^^
   |
   = note: creating a reference requires the pointer to be valid and imposes aliasing requirements
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
   |
LL |     let _p: *const i32 = addr_of!(*(&**w));
   |                                    +++  +

warning: 9 warnings emitted

